### ISSUE TITLE
Add SGD with Momentum

### ISSUE THREAD

**Issue by leonardoaraujosantos** [leonardoaraujosantos/DLMatFramework#12]
## Introduction
The Stochastic Gradient descent with Momentum normally perform better on deep learning models than the simple stochastic gradient descent.

## Physical interpretation
The loss can be interpreted as a the height of a high terrain. Initializing the parameters with random numbers is equivalent to setting a particle with zero initial velocity at some location. The optimization process can then be seen as equivalent to the process of simulating the parameter vector (i.e. a particle) as rolling on the landscape.
![saddle_point_evaluation_optimizers](https://cloud.githubusercontent.com/assets/898383/23307459/8cdee982-fa9f-11e6-810e-d9139227d505.gif)

The idea is that the SGD with momentum will build energy while going down on the loss hill, so it will get unstuck from local minimas.
![sgdmomentum](https://cloud.githubusercontent.com/assets/898383/23307328/fc58bd84-fa9e-11e6-8ab6-74c1fafe6b80.gif)

## Comparison between Vanilla SGD and SGD with momentum
``` python
# Vanilla update
x += - learning_rate * dx
```

``` python
# Momentum update
v = mu * v - learning_rate * dx # integrate velocity
x += v # integrate position
```
Here we have:
* mu: Parameter that control the friction of the particle the idea is to avoid the particle to build too much energy (Normally 0.9)
* v: Variable initialized with zero, v will represent the velocity of the particle

## Important point
Notice that each velocity vector (v) is going to be private to each layer parameter, so for example the FC layer will have one vector for weight and bias.

#### SGD Vanilla 
![sgd_vanilla](https://cloud.githubusercontent.com/assets/898383/23310486/9e9524b0-faaa-11e6-8f08-bfbd96fb9165.png)

#### With Momentum
![sgd_momentum](https://cloud.githubusercontent.com/assets/898383/23310264/e29a2f6c-faa9-11e6-9a29-214eb3a2b96e.png)

## References
* http://cs231n.github.io/neural-networks-3/#sgd
* http://www.yaldex.com/game-development/1592730043_ch18lev1sec4.html

**Comment by leonardoaraujosantos** [comment]
Still need to improve structure for other methods (ex Adam), but it's working... And as expected it converge faster than vanilla SGD
